File: Recursion/staircase.cpp
#include <bits/stdc++.h>
using namespace std;
vector<vector<int>> getWays(int cur, int n)
{
    if (cur > n)
        return {};
    if (cur == n)
        return {{}};
    auto ans1 = getWays(cur + 1, n);
    auto ans2 = getWays(cur + 2, n);
    vector<vector<int>> ans;
    for (auto arr : ans1)
    {
        arr.insert(arr.begin(), 1);
        ans.push_back(arr);
    }
    for (auto arr : ans2)

    {
        arr.insert(arr.begin(), 2);
        ans.push_back(arr);
    }
    return ans;
}
int main()
{
    int n = 10;
    vector<vector<int>> ways = getWays(0, n);
    for (auto &way : ways)
    {
        for (int step : way)
        {
            cout << step << " ";
        }
        cout << endl;
    }
}

File: Recursion/staircase.java
import java.util.*;

public class staircase{
    public static List<List<Integer>> getWays(int cur, int n) {
        if (cur > n)
            return new ArrayList<>();
        if (cur == n) {
            List<List<Integer>> baseCase = new ArrayList<>();
            baseCase.add(new ArrayList<>());
            return baseCase;
        }
        
        List<List<Integer>> ans1 = getWays(cur + 1, n);
        List<List<Integer>> ans2 = getWays(cur + 2, n);
        List<List<Integer>> ans = new ArrayList<>();
        
        for (List<Integer> arr : ans1) {
            List<Integer> temp = new ArrayList<>(arr);
            temp.add(0, 1);
            ans.add(temp);
        }
        for (List<Integer> arr : ans2) {
            List<Integer> temp = new ArrayList<>(arr);
            temp.add(0, 2);
            ans.add(temp);
        }
        return ans;
    }

    public static void main(String[] args) {
        int n = 10;
        List<List<Integer>> ways = getWays(0, n);
        
        for (List<Integer> way : ways) {
            for (int step : way) {
                System.out.print(step + " ");
            }
            System.out.println();
        }
    }
}